src/style-spec/expression/definitions/comparison.js

```diff
@@ -0,0 +1,170 @@
+ *
+ * For inequalities, T must be either value, string, or number. For ==/!=, it
+ * can also be boolean or null.
+ *
+ * Equality semantics are equivalent to Javascript's strict equality (===/!==)
+ * -- i.e., when the arguments' types don't match, == evaluates to false, != to
+ * true.
+ *
+ * When types don't match in an ordering comparison, a runtime error is thrown.
+ *
+ * @private
+ */
+function makeComparison(op: ComparisonOperator, compareBasic, compareWithCollator) {
+    return class Comparison implements Expression {
+        type: Type;
+        lhs: Expression;
+        rhs: Expression;
+        collator: ?Expression;
+        needsRuntimeTypeCheck: boolean;
+
+        constructor(lhs: Expression, rhs: Expression, collator: ?Expression) {
+            this.type = BooleanType;
+            this.lhs = lhs;
+            this.rhs = rhs;
+            this.collator = collator;
+            this.needsRuntimeTypeCheck = op !== '==' && op !== '!=' &&
+                lhs.type.kind === 'value' && rhs.type.kind === 'value';
+        }
+
+        static parse(args: Array<mixed>, context: ParsingContext): ?Expression {
+            if (args.length !== 3 && args.length !== 4)
+                return context.error(`Expected two or three arguments.`);
+
+            const op: ComparisonOperator = (args[0]: any);
+
+            let lhs = context.parse(args[1], 1, ValueType);
+            if (!lhs) return null;
+            if (!isComparableType(op, lhs.type)) {
+                return context.concat(1).error(`"${op}" comparisons are not supported for type '${toString(lhs.type)}'.`);
+            }
+            let rhs = context.parse(args[2], 2, ValueType);
+            if (!rhs) return null;
+            if (!isComparableType(op, rhs.type)) {
+                return context.concat(2).error(`"${op}" comparisons are not supported for type '${toString(rhs.type)}'.`);
+            }
+
+            if (!(
+                lhs.type.kind === rhs.type.kind ||
+                lhs.type.kind === 'value' ||
+                rhs.type.kind === 'value'
+            )) {
+                return context.error(`Cannot compare types '${toString(lhs.type)}' and '${toString(rhs.type)}'.`);
+            }
+
+            if (op !== '==' && op !== '!=') {
+                // typing rules specific to less/greater than operators
+                if (lhs.type.kind === 'value' && rhs.type.kind !== 'value') {
+                    // (value, T)
+                    lhs = new Assertion(rhs.type, [lhs]);
+                } else if (lhs.type.kind !== 'value' && rhs.type.kind === 'value') {
+                    // (T, value)
+                    rhs = new Assertion(lhs.type, [rhs]);
+                }
+            }
+
+            let collator = null;
+            if (args.length === 4) {
+                if (lhs.type.kind !== 'string' && rhs.type.kind !== 'string') {
+                    return context.error(`Cannot use collator to compare non-string types.`);
```

Comment:
Oh, good point

Reply:
🤔 if there's a collator argument, what should `["==", ["get", "name"], ["get", "name_en"], ["collator", {...}]]` do on a feature where one or both of `name` or `name_en` aren't strings?

Reply:
Two possibilities:
 - When there's a collator, infer `["string", ...]` annotations, so that it's a runtime error if either one is not a string. (So: no relaxed type semantics when there's a collator)
 - Use `collator.compare()` if both arguments evaluate to strings, and `===` otherwise.

Reply:
@nickidlugash ^ any gut feeling on this issue?

Reply:
@anandthakker For the use case of `["==", ["get", "name"], ["get", "name_en"], ["collator", {...}]]`, it will be common for there to be `null` values in either/both string. Given that, I think the desired behavior would be for this expression to return `false` if there is a `null` value, rather than producing an error. I think either the second or third option would give the desired result?

Are we also special-casing `null` values for other comparisons? E.g. does `["==", ["get", "name"], ["get", "name_en"]]` return `false` if one of the values is `null`? What about if the comparison was between a number and a `null` value?

Reply:
> Are we also special-casing null values for other comparisons? E.g. does ["==", ["get", "name"], ["get", "name_en"]] return false if one of the values is null? What about if the comparison was between a number and a null value?

Yep, these will both now return `false`